sw/device/tests/aes_sideload_test.c

```diff
@@ -0,0 +1,255 @@
+
+  keymgr_testutils_advance_state(&keymgr, &kOwnerIntParams);
+  keymgr_testutils_check_state(&keymgr, kDifKeymgrStateOwnerIntermediateKey);
+  LOG_INFO("Keymgr entered OwnerIntKey State");
+
+  keymgr_testutils_generate_identity(&keymgr);
+  LOG_INFO("Keymgr generated identity at OwnerIntKey State");
+
+  keymgr_testutils_generate_versioned_key(&keymgr, kKeyVersionedParams);
+  LOG_INFO("Keymgr generated SW output at OwnerIntKey State");
+
+  dif_keymgr_versioned_key_params_t sideload_params = kKeyVersionedParams;
+  sideload_params.dest = kDifKeymgrVersionedKeyDestAes;
+  keymgr_testutils_generate_versioned_key(&keymgr, sideload_params);
+  LOG_INFO("Keymgr generated HW output for Aes at OwnerIntKey State");
+}
+
+void aes_test(void) {
+  dif_aes_t aes;
+  CHECK_DIF_OK(
+      dif_aes_init(mmio_region_from_addr(TOP_EARLGREY_AES_BASE_ADDR), &aes));
+  CHECK_DIF_OK(dif_aes_reset(&aes));
+
+  // Setup ECB encryption transaction.
+  dif_aes_transaction_t transaction = {
+      .operation = kDifAesOperationEncrypt,
+      .mode = kDifAesModeEcb,
+      .key_len = kDifAesKey256,
+      .manual_operation = kDifAesManualOperationManual,
+      .masking = kDifAesMaskingInternalPrng,
+      .key_provider = kDifAesKeySideload,
+      .mask_reseeding = kDifAesReseedPer64Block,
+  };
+
+  CHECK_DIF_OK(dif_aes_start(&aes, &transaction, NULL, NULL));
+  dif_aes_data_t in_data_plain;
+  memcpy(in_data_plain.data, kPlainText, sizeof(kPlainText));
+  AES_TESTUTILS_WAIT_FOR_STATUS(&aes, kDifAesStatusInputReady, true, TIMEOUT);
+  CHECK_DIF_OK(dif_aes_load_data(&aes, in_data_plain));
+
+  // Trigger the AES encryption and wait for it to complete.
+  CHECK_DIF_OK(dif_aes_trigger(&aes, kDifAesTriggerStart));
+  AES_TESTUTILS_WAIT_FOR_STATUS(&aes, kDifAesStatusOutputValid, true, TIMEOUT);
+
+  //  Verify that the ciphertext is different from the plaintext.
+  dif_aes_data_t out_data;
+  CHECK_DIF_OK(dif_aes_read_output(&aes, &out_data));
+  CHECK_ARRAYS_NE(out_data.data, kPlainText, ARRAYSIZE(kPlainText));
+
+  // Finish the ECB encryption transaction.
+  CHECK_DIF_OK(dif_aes_end(&aes));
+
+  transaction.operation = kDifAesOperationDecrypt;
+  CHECK_DIF_OK(dif_aes_start(&aes, &transaction, NULL, NULL));
+  memcpy(in_data_plain.data, out_data.data, sizeof(out_data.data));
+  AES_TESTUTILS_WAIT_FOR_STATUS(&aes, kDifAesStatusInputReady, true, TIMEOUT);
+  CHECK_DIF_OK(dif_aes_load_data(&aes, in_data_plain));
+
+  // Trigger the AES decryption and wait for it to complete.
+  CHECK_DIF_OK(dif_aes_trigger(&aes, kDifAesTriggerStart));
+  AES_TESTUTILS_WAIT_FOR_STATUS(&aes, kDifAesStatusOutputValid, true, TIMEOUT);
+
+  // Verify that the output is equal to the plain text.
+  dif_aes_data_t out_data_plain;
+  CHECK_DIF_OK(dif_aes_read_output(&aes, &out_data_plain));
+  CHECK_ARRAYS_EQ(out_data_plain.data, kPlainText, ARRAYSIZE(kPlainText));
+
+  // Finish the ECB decryption transaction.
+  CHECK_DIF_OK(dif_aes_end(&aes));
```

Comment:
It would be good to also verify the key clear behavior by triggering `KEYMGR_SIDELOAD_CLEAR` and ensuring that AES gets either a garbled key or fails to process (I am not sure which of the two behaviors is implemented)

@tjaychen can provide more details if needed.

Reply:
Done, please check whether there's a better way to do it.

Reply:
Note: in manual operation mode (`manual_operation = kDifAesManualOperationManual`) AES will still start and use whatever bits keymgr provides. This mode is primarily intended for testing.

In automatic mode (`manual_operation = kDifAesManualOperationAuto`), AES will not start unless keymgr drives the valid bit high. This mode is intended by for real applications.

Using the manual mode for this test is still okay in my opinion. We do test the correct behavior of AES in both modes at the IP level.